Allow `PlanNode` to have multiple children

leiysky opened the pull request:
## Summary

So far, `fusequery` has strong assumption that a `PlanNode` only have one child(see `PlanNode::input()` and `PlanNode::set_input`).

To support operator with multiple children(e.g. `JOIN`, `UNION`), we should provide interface to access children nodes.

## Changelog

- Improvement

## Related Issues

## Test Plan

Unit Tests
Stateless Tests

BohuTANG commented:
@leiysky 
Hi, thanks for the contributing.
But `child` is a bit confusing here, for a pipeline(build from plan), it has two ports: input(s) and output(s) with direction hints, `child` more emphasis on hierarchical relationships.

@leiysky @zhang2014 
what do you think?

leiysky commented:
> @leiysky
> Hi, thanks for the contributing.
> But `child` is a bit confusing here, for a pipeline(build from plan), it has two ports: input(s) and output(s) with direction hints, `child` more emphasis on hierarchical relationships.
> 
> @leiysky @zhang2014
> what do you think?

It's okay to keep `input`.

IMHO, `PlanNode` doesn't need to consider about the direction of dataflow, instead they should be decoupled so we can replace the implementation easier.

Besides, a `PlanNode` is a **tree** of relational operators, for which `child` can describe the topology relationship between nodes better.

jyizheng commented:
> > @leiysky
> > Hi, thanks for the contributing.
> > But `child` is a bit confusing here, for a pipeline(build from plan), it has two ports: input(s) and output(s) with direction hints, `child` more emphasis on hierarchical relationships.
> > @leiysky @zhang2014
> > what do you think?
> 
> It's okay to keep `input`.
> 
> IMHO, `PlanNode` doesn't need to consider about the direction of dataflow, instead they should be decoupled so we can replace the implementation easier.
> 
> Besides, a `PlanNode` is a **tree** of relational operators, for which `child` can describe the topology relationship between nodes better.

I find something that might be relevant to the discussion: https://github.com/datafuselabs/datafuse/pull/563#issuecomment-842476092

BohuTANG commented:
> > @leiysky
> > Hi, thanks for the contributing.
> > But `child` is a bit confusing here, for a pipeline(build from plan), it has two ports: input(s) and output(s) with direction hints, `child` more emphasis on hierarchical relationships.
> > @leiysky @zhang2014
> > what do you think?
> 
> It's okay to keep `input`.
> 
> IMHO, `PlanNode` doesn't need to consider about the direction of dataflow, instead they should be decoupled so we can replace the implementation easier.
> 
> Besides, a `PlanNode` is a **tree** of relational operators, for which `child` can describe the topology relationship between nodes better.


> > @leiysky
> > Hi, thanks for the contributing.
> > But `child` is a bit confusing here, for a pipeline(build from plan), it has two ports: input(s) and output(s) with direction hints, `child` more emphasis on hierarchical relationships.
> > @leiysky @zhang2014
> > what do you think?
> 
> It's okay to keep `input`.
> 
> IMHO, `PlanNode` doesn't need to consider about the direction of dataflow, instead they should be decoupled so we can replace the implementation easier.
> 
> Besides, a `PlanNode` is a **tree** of relational operators, for which `child` can describe the topology relationship between nodes better.

OK.
Transform is directly and build from plan, input would be better to understand here(post-order/pre-order children confused?).
By the way, for join we need a JoinPlan with left and right inputs is ok, and no need all the plan has multiple inputs.Like datafusion did:
https://github.com/apache/arrow-datafusion/blob/master/datafusion/src/logical_plan/plan.rs#L106

leiysky commented:
@BohuTANG Yes, but we still need the `PlanNode::inputs` interface to access the children. See https://github.com/apache/arrow-datafusion/blob/6c050b896215717233dafe441b5c08ae40baf174/datafusion/src/logical_plan/plan.rs#L320-L339.

Also, I think the `PlanNode::set_input` interface can be removed since we usually modify a `PlanNode` with concrete variant.

BohuTANG commented:
> @BohuTANG Yes, but we still need the `PlanNode::inputs` interface to access the children. See https://github.com/apache/arrow-datafusion/blob/6c050b896215717233dafe441b5c08ae40baf174/datafusion/src/logical_plan/plan.rs#L320-L339.
> 
> Also, I think the `PlanNode::set_input` interface can be removed since we usually modify a `PlanNode` with concrete variant.

`inputs` and `set_input` can be removed  make sense.

leiysky commented:
@BohuTANG  I introduced a brand new `PlanRewriter` trait in this PR and refactored some code with it. Thus the original `PlanRewriter` struct is renamed to `RewriteHelper`.

@sundy-li As mentioned in #574 , there are only 1 test case broken by this PR, which is [test_select_interpreter](hhttps://github.com/datafuselabs/datafuse/blob/2b2881d8b1f20e86b531516c87f6380ba51a31a2/fusequery/query/src/interpreters/interpreter_select_test.rs. The cause is that after removing `AliasPushDownOptimizer`, fusequery doesn't have ability to infer field names, which I think can be fixed with your work #550 . So I just comment it out in this PR.

BohuTANG commented:
Really cool, looks good to me 🚀

leiysky commented:
@BohuTANG PTAL

BohuTANG commented:
> @BohuTANG PTAL

Nice, I will review it carefully.
Thank you @leiysky

BohuTANG commented on common/planners/src/plan_rewriter.rs:
````diff
@@ -4,16 +4,206 @@
+/// let mut rewriter = FilterRewriter {};
+/// let new_plan = rewriter.rewrite_plan_node(&plan)?; // new_plan is the rewritten plan
+/// ```
+pub trait PlanRewriter<'plan> {
````
This file is great, clear and neat.

BohuTANG commented on fusequery/query/src/planners/plan_scheduler.rs:
```diff
@@ -110,21 +108,26 @@ impl PlanScheduler {
-                let mut rewritten_node = PlanNode::Empty(EmptyPlan {
-                    schema: Arc::new(DataSchema::empty())
-                });
+                struct ReadSourceRewriter {
```
Can we move this block to out, seems better to read?

leiysky replied:
That's fine.

leiysky replied:
Done.
